File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/gui/activities/main/fragments/aurora_factors/AuroraFactorFragment.java
package se.gustavkarlsson.aurora_notifier.android.gui.activities.main.fragments.aurora_factors;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import se.gustavkarlsson.aurora_notifier.android.R;
import se.gustavkarlsson.aurora_notifier.android.gui.AuroraReportUpdateListener;
import se.gustavkarlsson.aurora_notifier.android.models.AuroraFactors;
import se.gustavkarlsson.aurora_notifier.android.models.AuroraReport;

public class AuroraFactorFragment extends Fragment implements AuroraReportUpdateListener {
	private static final String TAG = AuroraFactorFragment.class.getSimpleName();

	private View rootView;
	private GeomagActivityPresenter geomagActivityPresenter;
	private GeomagLocationPresenter geomagLocationPresenter;
	private WeatherPresenter weatherPresenter;
	private DarknessPresenter darknessPresenter;

	@Override
	public void onCreate(Bundle savedInstanceState) {
		Log.v(TAG, "onCreate");
		super.onCreate(savedInstanceState);
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		Log.v(TAG, "onCreateView");
		rootView = inflater.inflate(R.layout.fragment_aurora_factors, container, false);
		geomagActivityPresenter = new GeomagActivityPresenter((AuroraFactorView) rootView.findViewById(R.id.aurora_factor_geomag_activity));
		geomagLocationPresenter = new GeomagLocationPresenter((AuroraFactorView) rootView.findViewById(R.id.aurora_factor_geomag_location));
		weatherPresenter = new WeatherPresenter((AuroraFactorView) rootView.findViewById(R.id.aurora_factor_weather));
		darknessPresenter = new DarknessPresenter((AuroraFactorView) rootView.findViewById(R.id.aurora_factor_darkness));
		return rootView;
	}

	@Override
	public void onUpdate(AuroraReport report) {
		AuroraFactors factors = report.getFactors();
		geomagActivityPresenter.onUpdate(factors.getGeomagActivity());
		geomagLocationPresenter.onUpdate(factors.getGeomagLocation());
		weatherPresenter.onUpdate(factors.getWeather());
		darknessPresenter.onUpdate(factors.getDarkness());
		rootView.invalidate();
	}
}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/gui/activities/main/fragments/aurora_factors/AuroraFactorFragment.kt
package se.gustavkarlsson.skylight.android.gui.activities.main.fragments.aurora_factors

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jakewharton.rxbinding2.view.clicks
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.fragment_aurora_factors.*
import org.jetbrains.anko.alert
import org.jetbrains.anko.okButton
import se.gustavkarlsson.skylight.android.R
import se.gustavkarlsson.skylight.android.appComponent
import se.gustavkarlsson.skylight.android.gui.AutoDisposingFragment
import timber.log.Timber

class AuroraFactorFragment : AutoDisposingFragment() {

	private val darknessViewModel: DarknessViewModel by lazy {
		appComponent.darknessViewModel(this)
	}

	private val geomagLocationViewModel: GeomagLocationViewModel by lazy {
		appComponent.geomagLocationViewModel(this)
	}

	private val kpIndexViewModel: KpIndexViewModel by lazy {
		appComponent.kpIndexViewModel(this)
	}

	private val visibilityViewModel: VisibilityViewModel by lazy {
		appComponent.visibilityViewModel(this)
	}

	override fun onCreateView(
		inflater: LayoutInflater,
		container: ViewGroup?,
		savedInstanceState: Bundle?
	): View {
		return inflater.inflate(R.layout.fragment_aurora_factors, container)
	}

	override fun onStart() {
		super.onStart()
		bindData()
	}

	private fun bindData() {
		bindFactor(
			darknessViewModel, darkness,
			R.string.factor_darkness_title_full, R.string.factor_darkness_desc, "darkness"
		)
		bindFactor(
			geomagLocationViewModel,
			geomagLocation,
			R.string.factor_geomag_location_title_full,
			R.string.factor_geomag_location_desc,
			"geomagLocation"
		)
		bindFactor(
			kpIndexViewModel, kpIndex,
			R.string.factor_kp_index_title_full, R.string.factor_kp_index_desc, "kpIndex"
		)
		bindFactor(
			visibilityViewModel, visibility,
			R.string.factor_visibility_title_full, R.string.factor_visibility_desc, "visibility"
		)
	}

	private fun bindFactor(
		viewModel: FactorViewModel<*>,
		view: AuroraFactorView,
		titleResourceId: Int,
		descriptionResourceId: Int,
		factorDebugName: String
	) {
		viewModel.value
			.doOnNext { Timber.d("Updating %s value view: %s", factorDebugName, it) }
			.observeOn(AndroidSchedulers.mainThread())
			.subscribe { view.value = it }
			.autoDisposeOnStop()

		viewModel.chance
			.doOnNext { Timber.d("Updating %s chance view: %s", factorDebugName, it) }
			.observeOn(AndroidSchedulers.mainThread())
			.subscribe { view.chance = it }
			.autoDisposeOnStop()

		view.clicks()
			.observeOn(AndroidSchedulers.mainThread())
			.subscribe {
				toastFactorInfo(titleResourceId, descriptionResourceId)
			}
			.autoDisposeOnStop()
	}

	private fun toastFactorInfo(
		titleResourceId: Int,
		descriptionResourceId: Int
	) {
		context?.alert {
			iconResource = R.drawable.info_white_24dp
			title = ctx.getString(titleResourceId)
			message = ctx.getString(descriptionResourceId)
			okButton { it.dismiss() }
		}?.show()
	}
}
